### PR TITLE
Made HardFork action drop all other actions

### LINKED ISSUE #3693: Discard all proposals upon HardForkInitiation ratification.
Upon ratification of a new HardForkInitiation governance action we must discard all of the governance proposals that are currently being considered. 

This is necessary because some of the proposals will become invalidated upon transition to a new era, a good example from the past would be removal of a protocol parameter.

Moreover, HardFork does not get Enacted until one epoch after being ratified, therefore we need to also prevent any new proposal procedures being accepted.

```
                    .--Epoch of no new Proposals
                    |
E0----------E1------^-----E2-------------E3
             |             |
             `RatifyHF     `EnactHF
               |              |
      (drop all Proposals)   (start accepting new Proposals)
```

### PR DESCRIPTION
**Pull request by Soupstraw**
# Description

Fixes #3693 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Arbitrary.hs [hunk 1]
```diff
@@ -127,6 +127,13 @@ instance
       <$> arbitrary
       <*> arbitrary
 
+instance Arbitrary RatifyStrategy where
+  arbitrary = elements
+    [ Continue
+    , Delay
+    , Drop
+    ]
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Arbitrary.hs]: Derive Enum and Bounded and you can use this implementation:
```suggestion
  arbitrary = arbitraryBoundedEnum
  shrink = shrinkBoundedEnum
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs [hunk 2]
```diff
@@ -198,14 +199,14 @@ dRepAcceptedRatio RatifyEnv {reDRepDistr, reDRepState, reCurrentEpoch} gasDRepVo
...
-delayingAction TreasuryWithdrawals {} = False
-delayingAction ParameterChange {} = False
-delayingAction InfoAction {} = False
+actionFutureStrategy :: GovAction era -> RatifyStrategy
+actionFutureStrategy NoConfidence {} = Continue
+actionFutureStrategy NewCommittee {} = Continue
+actionFutureStrategy NewConstitution {} = Continue
+actionFutureStrategy TreasuryWithdrawals {} = Delay
+actionFutureStrategy ParameterChange {} = Delay
+actionFutureStrategy InfoAction {} = Delay
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: I think you mixed them up
```suggestion
actionFutureStrategy NoConfidence {} = Delay
actionFutureStrategy NewCommittee {} = Delay
actionFutureStrategy NewConstitution {} = Delay
actionFutureStrategy TreasuryWithdrawals {} = Continue
actionFutureStrategy ParameterChange {} = Continue
actionFutureStrategy InfoAction {} = Continue
```

**Review by lehins — CHANGES_REQUESTED** [review]
Except one potential mistake, this looks good.

**Comment by Soupstraw** [comment]
I'm thinking this wouldn't probably work the way I did it. We also somehow have to prevent adding new GAs after HF is ratified, probably by storing the RatifyStrategy for future epochs.

**Comment by lehins** [comment]
> I'm thinking this wouldn't probably work the way I did it. We also somehow have to prevent adding new GAs after HF is ratified, probably by storing the RatifyStrategy for future epochs.

@Soupstraw It would work, you just haven't implemented it all the way

We need another flag in the enact state that would prevent new proposals whenever Ratify ended with a Drop strategy

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs (hunk 2 above)
Comment (Soupstraw) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: nice catch!

**Comment by lehins** [comment]
I think it will be better to clear out all of the proposals during translation, which will be implemented in the next era after Conway.

It would make a little more sense if we were leaning towards delaying for two epochs, but since now this is changing, dropping proposals at the same time we do a hardfork makes the rule a bit more complicated than it should be.

**Pull request closed without merging**